linkerd/distribute/src/params.rs

```diff
@@ -16,17 +20,38 @@ pub enum Distribution<K> {
     Empty,
 
     /// A distribution that uses the first available backend in an ordered list.
-    FirstAvailable(Arc<[K]>),
+    FirstAvailable(Arc<UnweightedKeys<K>>),
 
     /// A distribution that uses the first available backend when randomly
     /// selecting over a weighted distribution of backends.
     RandomAvailable(Arc<WeightedKeys<K>>),
 }
 
+// Uniquely identifies a key/backend pair for a distribution. This allows
+// backends to have the same key and still participate in request distribution.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct KeyId {
+    idx: usize,
+}
+
+impl KeyId {
+    pub(crate) fn new(idx: usize) -> Self {
+        Self { idx }
+    }
+}
+
+#[derive(Debug)]
+pub struct UnweightedKeys<K> {
+    ids: Vec<KeyId>,
```

Comment:
This is the key part of this change. It decouples the backend keys from the iteration order of the backends with a separate key type. I considered using something like [SlotMap](https://docs.rs/slotmap/latest/slotmap/), but since we don't modify or remove anything from these maps after they're created we don't need the guarantee of stable keys that it provides and the slight performance cost that comes with it.